### PR TITLE
serialize to nested tibble

### LINKED ISSUE #595: Serialize to/from nested tibble
Based on a "root table", with `nest_join()` (#282) and `pack_join()` (TBD).

The serialization to a tibble will, in general, produce redundancies. No redundancies if all tables are either detail or parent tables of the root table (directly or indirectly).

For serialization from a tibble, we may want to deduplicate those redundancies.

Application: JSON objects often map to a tibble with nested + packed columns, this will offer a way to serialize to/from JSON.

### PR DESCRIPTION
**Pull request by moodymudskipper**
Ultimately closes #595

4 functions make a round trip : `dm_to_tibble`, `serialize_list_cols`, `unserialize_json_cols` and `tibble_to_dm`.

Serialization is wrong at the moment.

### CODE REVIEW

**Comment by krlmlr** [comment]
Thanks. It works for me, I'm seeing a few problems:

``` r
library(tidyverse)
library(dm)

# x <- dm_to_tibble(dm_nycflights13(), flights)
x <- dm:::dm_to_tibble(dm_nycflights13(), "flights")
x
#> $tbl
#> # A tibble: 1,761 × 23
#>     year month   day dep_time sched_dep_time dep_delay arr_time sched_arr_time
#>    <int> <int> <int>    <int>          <int>     <dbl>    <int>          <int>
#>  1  2013     1    10        3           2359         4      426            437
#>  2  2013     1    10       16           2359        17      447            444
#>  3  2013     1    10      450            500       -10      634            648
#>  4  2013     1    10      520            525        -5      813            820
#>  5  2013     1    10      530            530         0      824            829
#>  6  2013     1    10      531            540        -9      832            850
#>  7  2013     1    10      535            540        -5     1015           1017
#>  8  2013     1    10      546            600       -14      645            709
#>  9  2013     1    10      549            600       -11      652            724
#> 10  2013     1    10      550            600       -10      649            703
#> # … with 1,751 more rows, and 15 more variables: arr_delay <dbl>,
#> #   carrier <chr>, flight <int>, tailnum <chr>, origin <chr>, dest <chr>,
#> #   air_time <dbl>, distance <dbl>, hour <dbl>, minute <dbl>, time_hour <dttm>,
#> #   airlines <tibble[,1]>, airports <tibble[,7]>, planes <tibble[,8]>,
#> #   weather <tibble[,13]>
#> 
#> $keys
#> $keys$own_keys
#> $keys$own_keys$fks
#> # A tibble: 4 × 5
#>   child_table child_fk_cols     parent_table parent_key_cols   on_delete
#>   <chr>       <keys>            <chr>        <keys>            <chr>    
#> 1 flights     carrier           airlines     carrier           no_action
#> 2 flights     origin            airports     faa               no_action
#> 3 flights     tailnum           planes       tailnum           no_action
#> 4 flights     origin, time_hour weather      origin, time_hour no_action
#> 
#> 
#> $keys$parents_keys
#> $keys$parents_keys$airlines
#> $keys$parents_keys$airlines$own_keys
#> $keys$parents_keys$airlines$own_keys$pks
#> [1] "carrier"
#> 
#> 
#> 
#> $keys$parents_keys$airports
#> $keys$parents_keys$airports$own_keys
#> $keys$parents_keys$airports$own_keys$pks
#> [1] "faa"
#> 
#> 
#> 
#> $keys$parents_keys$planes
#> $keys$parents_keys$planes$own_keys
#> $keys$parents_keys$planes$own_keys$pks
#> [1] "tailnum"
#> 
#> 
#> 
#> $keys$parents_keys$weather
#> $keys$parents_keys$weather$own_keys
#> $keys$parents_keys$weather$own_keys$pks
#> [1] "origin"    "time_hour"

dm <- dm:::tibble_to_dm(x$tbl, x$keys, "flights")

# Problem: primary key of flights table
dm %>%
  dm_paste()
#> dm::dm(
#>   airlines,
#>   airports,
#>   planes,
#>   weather,
#>   flights,
#> ) %>%
#>   dm::dm_add_pk(airlines, carrier) %>%
#>   dm::dm_add_pk(airports, faa) %>%
#>   dm::dm_add_pk(planes, tailnum) %>%
#>   dm::dm_add_pk(weather, c(origin, time_hour)) %>%
#>   dm::dm_add_pk(flights, c()) %>%
#>   dm::dm_add_fk(flights, carrier, airlines) %>%
#>   dm::dm_add_fk(flights, origin, airports) %>%
#>   dm::dm_add_fk(flights, tailnum, planes) %>%
#>   dm::dm_add_fk(flights, c(origin, time_hour), weather)

# Problem: rows in dimension tables should be deduplicated
dm %>%
  dm_rm_pk(flights) %>%
  dm_examine_constraints()
#> ! Unsatisfied constraints:
#> • Table `airlines`: primary key `carrier`: has duplicate values: B6 (296), UA (294), EV (278), DL (239), AA (180), …
#> • Table `airports`: primary key `faa`: has duplicate values: EWR (641), JFK (602), LGA (518)
#> • Table `planes`: primary key `tailnum`: has 9 missing values, and duplicate values: N197JB (6), N12142 (6), N14568 (6), N15980 (6), N16919 (6), N190JB (NA)
#> • Table `weather`: primary key `origin`   , `time_hour`: has duplicate values: EWR, 2013-01-10 06:00:00 (31), JFK, 2013-01-10 08:00:00 (30), JFK, 2013-02-10 08:00:00 (30), EWR, 2013-01-10 08:00:00 (29), EWR, 2013-01-10 07:00:00 (26), …
```

<sup>Created on 2022-01-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by krlmlr** [comment]
I'm still not in love with the `keys` data structure, and I think we do want to expose `dm_to_tibble()`. I'd like dm to consist of a set of composable tools.

How about we define two new verbs: `dm_unnest()` and `dm_unpack()`. Then, `dm_to_tibble()` prints how to convert the output back to a dm:

``` r
library(tidyverse)
library(dm)

dm_to_tibble(dm_nycflights13(), flights)
#> Convert to a dm with: %>%
#>   as_dm(name = "flights") %>%
#>   dm_unnest(flights, carrier, airlines) %>%
#>   dm_unnest(flights, origin, airports) %>%
#>   dm_unnest(flights, tailnum, planes) %>%
#>   dm_unnest(flights, c(origin, time_hour), weather)
#> # A tibble: 1,761 × 23
#>     year month   day dep_time sched_dep_time dep_delay arr_time sched_arr_time
#>    <int> <int> <int>    <int>          <int>     <dbl>    <int>          <int>
#>  1  2013     1    10        3           2359         4      426            437
#>  2  2013     1    10       16           2359        17      447            444
#>  3  2013     1    10      450            500       -10      634            648
#>  4  2013     1    10      520            525        -5      813            820
#>  5  2013     1    10      530            530         0      824            829
#>  6  2013     1    10      531            540        -9      832            850
#>  7  2013     1    10      535            540        -5     1015           1017
#>  8  2013     1    10      546            600       -14      645            709
#>  9  2013     1    10      549            600       -11      652            724
#> 10  2013     1    10      550            600       -10      649            703
#> # … with 1,751 more rows, and 15 more variables: arr_delay <dbl>,
#> #   carrier <chr>, flight <int>, tailnum <chr>, origin <chr>, dest <chr>,
#> #   air_time <dbl>, distance <dbl>, hour <dbl>, minute <dbl>, time_hour <dttm>,
#> #   airlines <tibble[,1]>, airports <tibble[,7]>, planes <tibble[,8]>,
#> #   weather <tibble[,13]>
```

<sup>Created on 2022-01-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

I'm not even sure we need two verbs -- we could have one `dm_unwrap()` that unnests list columns and unpacks data frame columns. Similarly, we might want to start with one `dm_wrap()` that runs one step of the process for one table, and turn `dm_to_tibble()` into a sequence of `dm_wrap()` calls internally.

File: tests/testthat/_snaps/dm_to_tibble.md [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+      1     4 streetlamp h     
+      
+
+---
+
+    Code
+      print(serialized_from_tf_4)
+    Output
+      # A tibble: 5 x 6
+        h     i     j        j1 tf_5_json_child             tf_3_json_parent
```
Comment (krlmlr) [inline tests/testthat/_snaps/dm_to_tibble.md:86]: Should we keep one column per parent or child table also in serialized form?

File: tests/testthat/_snaps/dm_to_tibble.md [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    Code
+      print(serialized_from_tf_4$tf_5_json_child)
+    Output
+      [1] "{\"empty\":[true],\"data\":[{\"k\":1,\"m\":\"a\",\"tf_6_json_parent\":\"{\\\"data\\\":[{}]}\"}]}"                        
+      [2] "{\"empty\":{},\"data\":[{\"k\":1,\"m\":\"house\",\"tf_6_json_parent\":\"{\\\"data\\\":[{\\\"o\\\":\\\"e\\\"}]}\"}]}"     
+      [3] "{\"empty\":{},\"data\":[{\"k\":2,\"m\":\"tree\",\"tf_6_json_parent\":\"{\\\"data\\\":[{\\\"o\\\":\\\"f\\\"}]}\"}]}"      
+      [4] "{\"empty\":{},\"data\":[{\"k\":3,\"m\":\"streetlamp\",\"tf_6_json_parent\":\"{\\\"data\\\":[{\\\"o\\\":\\\"h\\\"}]}\"}]}"
+      [5] "{\"empty\":{},\"data\":[{\"k\":4,\"m\":\"streetlamp\",\"tf_6_json_parent\":\"{\\\"data\\\":[{\\\"o\\\":\\\"h\\\"}]}\"}]}"
+    Code
+      print(serialized_from_tf_4$tf_3_json_parent)
```
Comment (krlmlr) [inline tests/testthat/_snaps/dm_to_tibble.md:102]: Can we use `as_glue()` for the column data (or at least for the printing)?

File: tests/testthat/_snaps/dm_to_tibble.md [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+       [2] "{\"empty\":{},\"data\":[{\"c\":\"elephant\",\"e\":\"D\",\"e1\":4,\"tf_3_json_parent\":\"{\\\"data\\\":[{\\\"g\\\":\\\"three\\\",\\\"tf_4_json_child\\\":\\\"{\\\\\\\"empty\\\\\\\":{},\\\\\\\"data\\\\\\\":[{\\\\\\\"h\\\\\\\":\\\\\\\"b\\\\\\\",\\\\\\\"i\\\\\\\":\\\\\\\"four\\\\\\\",\\\\\\\"tf_5_json_child\\\\\\\":\\\\\\\"{\\\\\\\\\\\\\\\"empty\\\\\\\\\\\\\\\":{},\\\\\\\\\\\\\\\"data\\\\\\\\\\\\\\\":[{\\\\\\\\\\\\\\\"k\\\\\\\\\\\\\\\":1,\\\\\\\\\\\\\\\"m\\\\\\\\\\\\\\\":\\\\\\\\\\\\\\\"house\\\\\\\\\\\\\\\",\\\\\\\\\\\\\\\"tf_6_json_parent\\\\\\\\\\\\\\\":\\\\\\\\\\\\\\\"{\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"data\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\":[{\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"o\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\":\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"e\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\"}]}\\\\\\\\\\\\\\\"}]}\\\\\\\"}]}\\\"}]}\"}]}"
```
Comment (krlmlr) [inline tests/testthat/_snaps/dm_to_tibble.md:382]: This looks like too many backslashes to me. Is this JSON packed inside JSON?

Comment (moodymudskipper) [reply tests/testthat/_snaps/dm_to_tibble.md:382]: Yes it is, so columns can be unwrapped recursively from SQL if needed. Isn't it what we want?

**Comment by moodymudskipper** [comment]
> I'm still not in love with the `keys` data structure, and I think we do want to expose `dm_to_tibble()`. I'd like dm to consist of a set of composable tools.

This looks good but we still need these keys somewhere don't we?
Now we have a 2 element list. Alternatives I see are : 
* (1) Make it an attribute to the table (at least now it'd be a single attribute at the top and not attributes to columns) 
* (2) collapse into a 1 table dm instead of 1 table, and store key info (and of course, the name of root table) in the dm
* (3) a new object type
* (4) store info in col names
* (5) we drop the keys and we try to guess

 I think I like (2), do you have other ideas?

**Comment by moodymudskipper** [comment]
`dm_wrap()` and `dm_unwrap()` might be the "smart" versions and then we have `dm_pack()`, `dm_unpack()`, `dm_nest()`, `dm_unnest()`, which might be better named `dm_pack_join()`, `dm_unpack_split()`, `dm_nest_join()`, `dm_unnest_split()`?

**Comment by moodymudskipper** [comment]
> I think I like (2)

I think we could keep the key structure of a regular dm and lose the nested structure I use, we might just flag the pks and fks as "deep" with a new column. When we wrap we flag relevant keys, when we unwrap we unflag. Validating functions either ignore flagged keys or unwrap all before checks.

**Comment by moodymudskipper** [comment]
> Thanks. It works for me, I'm seeing a few problems:

Now fixed : 

``` r
library(tidyverse)
library(dm, warn.conflicts = FALSE)

x <- dm:::dm_to_tibble(dm_nycflights13(), "flights")
dm <- dm:::tibble_to_dm(x$tbl, x$keys, "flights")

dm %>%
  dm_paste()
#> dm::dm(
#>   airlines,
#>   airports,
#>   planes,
#>   weather,
#>   flights,
#> ) %>%
#>   dm::dm_add_pk(airlines, carrier) %>%
#>   dm::dm_add_pk(airports, faa) %>%
#>   dm::dm_add_pk(planes, tailnum) %>%
#>   dm::dm_add_pk(weather, c(origin, time_hour)) %>%
#>   dm::dm_add_fk(flights, carrier, airlines) %>%
#>   dm::dm_add_fk(flights, origin, airports) %>%
#>   dm::dm_add_fk(flights, tailnum, planes) %>%
#>   dm::dm_add_fk(flights, c(origin, time_hour), weather)

dm %>%
  dm_examine_constraints()
#> ! Unsatisfied constraints:
#> • Table `flights`: foreign key `tailnum` into table `planes`: values of `flights$tailnum` not in `planes$tailnum`: N725MQ (6), N537MQ (5), N722MQ (5), N730MQ (5), N736MQ (5), …
```

<sup>Created on 2022-01-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

The unsatisfied constraint is the same as in the original dm

File: tests/testthat/_snaps/dm_to_tibble.md (hunk 1 above)
Comment (moodymudskipper) [reply tests/testthat/_snaps/dm_to_tibble.md:86]: I don't understand

File: tests/testthat/_snaps/dm_to_tibble.md (hunk 2 above)
Comment (moodymudskipper) [reply tests/testthat/_snaps/dm_to_tibble.md:102]: done

**Comment by krlmlr** [comment]
Looks good! Let's scope this PR to dm -> tibble -> dm and keep the functions unexported. Tests must pass, we can skip for databases. New issues follow, we tackle one by one.

**Comment by krlmlr** [comment]
Based on `dm_wrap()`.

**Pull request closed without merging**